server/src/main/java/org/elasticsearch/common/io/stream/DelayableWriteable.java

```diff
@@ -28,10 +28,23 @@
 import java.util.function.Supplier;
 
 /**
- * A holder for {@link Writeable}s that can delays reading the underlying
- * {@linkplain Writeable} when it is read from a remote node.
+ * A holder for {@link Writeable}s that delays reading the underlying object
+ * on the receiving end. To be used for objects that use a lot of memory hence
+ * it is desirable to keep them around only for a limited amount of time.
+ * The node that produces the {@link Writeable} calls {@link #referencing(Writeable)}
+ * to create a {@link DelayableWriteable} that serializes the inner object
+ * first to a buffer and writes the content of the buffer to the {@link StreamOutput}.
+ * The receiver node calls {@link #delayed(Reader, StreamInput)} to create a
+ * {@link DelayableWriteable} that reads the buffer from the @link {@link StreamInput}
+ * but delays creating the actual object by calling {@link #expand()} when needed.
+ * Multiple {@link DelayableWriteable}s coming from different nodes may be buffered
+ * on the receiver end, which may hold a mix of {@link DelayableWriteable}s that were
+ * produced locally (hence expanded) as well as received form another node (hence subject
+ * to delayed expansion). When such objects are buffered for some time it is desirable
+ * to force their buffering in serialized format by calling
+ * {@link #asSerialized(Reader, NamedWriteableRegistry)}.
  */
-public abstract class DelayableWriteable<T extends Writeable> implements Supplier<T>, Writeable {
+public abstract class DelayableWriteable<T extends Writeable> implements Writeable {
```

Comment:
I was wondering, is it just DelayedWriteable or DelayedReadWriteable ? Because it's not that you can choose to delay its reading, it's the only way that it works?

Reply:
I guess so. I figured `Writeable` was the name for things that we threw across the wire and calling it `Delayed` was enough to imply that it delayed reading because that was sort of the only thing you could delay.

At this point, given how we use it, it might be more correct to call it `PotentiallySerializedWriteable`. But that is kind of mouthful. I think the whole "Delayed" bit of the name was perfect for the first PR that built it but it is less and less perfect now.

Reply:
but why Delayed VS Delayable for instance?

Reply:
Because the `referencing` implementation isn't delayed.

Reply:
I find that that is a lie :) because writeTo will always do the two steps serialization that will allow for delaying the deserialization on the other side. The point is that referencing is only on the producer side hence you never read into referencing from the stream? Maybe a bit of a philosophical discussion.

Reply:
I think we should rename it to something like `SerializedReference` or `MaybeSerializedReference` so maybe its not an issue then. I get where you are coming from though. The name is icky now.